doc/userguide/rules/rules-internals.rst

```diff
@@ -0,0 +1,236 @@
+
+    pass tls $HOME_NET any -> $EXTERNAL_NET any (tls.sni; content:"some.\
+    content"; startswith; nocase; endswith; msg:"matching TLS allowed";\
+    priority:1; flow:to_server, established; content:"hypothetical"; sid:6;\
+    rev:1; gid:10000003;)
+
+With the default `action order configuration <suricata-yaml-action-order>`_
+(*pass*, *drop*, *reject*, *alert*), after loading and sorting, their
+internal IDs would be:
+
++--------------------+-------------------+
+| Signature ID (sid) | Internal ID (iid) |
++====================+===================+
+|  1                 |  3                |
++--------------------+-------------------+
+|  2                 |  0                |
++--------------------+-------------------+
+|  3                 |  5                |
++--------------------+-------------------+
+|  4                 |  4                |
++--------------------+-------------------+
+|  5                 |  1                |
++--------------------+-------------------+
+|  6                 |  2                |
++--------------------+-------------------+
+
+
+Inspection Flow
+---------------
+
+Once it is time to inspect network traffic against the loaded rules, the
+Detect Engine will match against - if applicable:
+
+#. IP Only rules;
+#. Packet/payload-related rules;
+#. Frame keywords;
+#. Application layer protocol transaction rules.
+
+During packet inspection, if the signature uses the last two in this list,
+inspection is left to those steps.
+
+.. tip::
+
+  With the introduction of :doc:`Firewall mode <../firewall/firewall-design>`,
+  it is possible to explicitly control to which step of the detection engine flow
+  a rule will be hooked. This is done with :ref:`Explicit rule hooks <rule-hooks>`.
+
+Each rule that matches becomes a ``PacketAlert``. After all matches for a
+packet have been processed, and the :ref:`alert queue limit <alert queue
+overflow impact>` is taken into account, ``PacketAlerts`` become the alerts
+that are seen in Suricata logs.
+
+Considerations on inspection steps
+----------------------------------
+
+IP Only rules
+~~~~~~~~~~~~~
+
+Without optimization, IP Only signatures would match on every packet on a flow.
+To improve performance, what Suricata does is to evaluate rules that are
+``ip-only`` only once per flow direction, for the first packet in each direction.
+
+Application layer protocol transactions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Each parser has its own state machine, where it uses a per-direction parsing
+state "progress". For each progress value keywords can be registered. So, for
+instance, ``http`` has a value "request line available" for which there are
+keywords like :ref:`http.uri <rules-http-uri-normalization>`, :ref:`http.method
+<http.method>` etc. registered. If in the traffic the engine reaches this state,
+the signatures with those keywords may be already evaluated, even if they have a
+lower priority than an http body inspecting signature.
```

Comment:
You may want to mention somehow that a rule with 2 keywords at different progress may be evaluated at 2 different packets